osquery/filesystem/fileops.h

```
@@ -0,0 +1,211 @@
+};
+
+/// Constant for an invalid handle
+const PlatformHandle kInvalidHandle = (PlatformHandle)-1;
+
+/**
+ * @brief File access modes for PlatformFile
+ *
+ * A file can be opened for many access modes with a variety of different
+ * options on Windows and POSIX. To provide multi-platform support, we need to
+ * provide an abstraction that can cover the supported platforms.
+ */
+
+#define PF_READ 0x0001
+#define PF_WRITE 0x0002
+
+#define PF_OPTIONS_MASK 0x001c
+#define PF_GET_OPTIONS(x) ((x & PF_OPTIONS_MASK) >> 2)
+#define PF_CREATE_NEW (0 << 2)
+#define PF_CREATE_ALWAYS (1 << 2)
+#define PF_OPEN_EXISTING (2 << 2)
+#define PF_TRUNCATE (3 << 2)
+
+#define PF_NONBLOCK 0x0020
+
+/**
+ * @brief Modes for seeking through a file
+ *
+ * Provides a chance 
+ */
+
+enum SeekMode {
+  PF_SEEK_BEGIN = 0,
+  PF_SEEK_CURRENT,
+  PF_SEEK_END
+};
+
+#ifdef WIN32
+
+/**
+ * @brief Windows-only class that deals with simulating POSIX 
+ *        asynchronous IO semantics using Windows API calls
+ */
+struct AsyncEvent {
+  AsyncEvent();
+  ~AsyncEvent();
+ 
+  OVERLAPPED overlapped_{ 0 };
+  std::unique_ptr<char[]> buffer_{ nullptr };
+  bool is_active_{ false };
+};
+
+#endif
+
+/**
+ * @brief Platform-agnostic file object
+ *
+ * PlatformFile is a multi-platform class that offers input/output capabilities for files. 
+ */
+class PlatformFile {
+  public:
+    explicit PlatformFile(const std::string& path, int mode, int perms = -1);
+    explicit PlatformFile(PlatformHandle handle) : handle_(handle) {}
+
+    PlatformFile(PlatformFile&& src) {
+      handle_ = kInvalidHandle;
+      std::swap(handle_, src.handle_);
+    }
+
+    ~PlatformFile();
+
+    /// Checks to see if the file object is actually a file and not a "special file"
+    bool isFile() const;
+
+    bool hasPendingIo() const { return has_pending_io_;  }
+    bool isValid() const { return (handle_ != kInvalidHandle); }
+    PlatformHandle nativeHandle() const { return handle_; }
+
+    bool isOwnerRoot() const;
```

What should the error response be? If `fstat` fails, is it better to throw an exception instead? Turning it into a `getOwner` type function will require mapping Windows accounts into `uid`s which is going to be difficult (for example, who is considered `root`?).